### ISSUE TITLE
[🐛 BUG]: invalid socket DSN (tcp://:6001, unix://file.sock) while trying to run rr workers and rr reset commands

### ISSUE THREAD

**Issue by DigitVE** [roadrunner-server/roadrunner#1610]
### No duplicates 🥲.

- [X] I have searched for a similar issue in our bug tracker and didn't find any solutions.

### What happened?

Hello! I can’t run commands `rr workers` and `rr reset` on prod environment (runs on Kubernetes) because of following error:
```
invalid socket DSN (tcp://:6001, unix://file.sock)
```
But in case if I use `rr workers -o rpc.listen=tcp://localhost:6001` it works just fine. There was also an attempt to use `rr workers -c /app/.rr.yaml` but unfortunately error still remained with the -c argument as well.

Unfortunately, I can’t share any deployment configs here because of security reasons but I would really appreciate any idea that hint me what I should check to solve this problem.

### Version (rr --version)

2.8.0 and higher, on lower versions it works ok

### How to reproduce the issue?

There is no way to reproduce it because locally it works without any problems

**Comment by rustatian** [comment]
Hey @DigitVE 👋🏻 
You're on the very-very-very old RR version. Consider updating to the recent one: `v2023.1.5`.

EDIT: Please, share your configuration as well.

**Comment by DigitVE** [comment]
@rustatian Thank you for your fast response! I've tried already on latest version (`v2023.1.5`) and bug remained there too.

Here is config without services and metrics:

```
version: "2.7"

rpc:
    listen: 'tcp://127.0.0.1:6001'

server:
    command: "php bin/console baldinof:roadrunner:worker"

http:
    address: 0.0.0.0:80
    middleware: ["static", "gzip", "http_metrics"]
    uploads:
        forbid: [".php", ".exe", ".bat"]
    static:
        dir: "public"
        forbid: [".php", ".htaccess"]
    pool:
        debug: false
        num_workers: 10
        max_jobs: 1000

logs:
    mode: production
    channels:
        http:
            level: info # Log all http requests, set to info to disable
        server:
            level: info # Everything written to worker stderr is logged
            mode: raw
        metrics:
            level: error
        service:
            level: info
            mode: raw

service:
...

metrics:
...

status:
    address: 0.0.0.0:2114
```

**Comment by rustatian** [comment]
`version: '2.7'` should be replaced by `version: '3'` for the `v2023.x.x`.

Tried right now: 
![image](https://github.com/roadrunner-server/roadrunner/assets/8040338/e0e8d927-b8fc-4664-bca1-863385db0c9b)

Everything looks good. Check the following:
1. By default (without `-c`) RR searches for the configuration in the `CWD` with the name `.rr.yaml`. Might be in that configuration you missed `rpc.listen` option.
2. Check the RR logs. Might be some info would be there.

**Comment by rustatian** [comment]
3. Check your production `.rr.yaml` for the `rpc.listen` option. It looks like it missed/corrupted.

**Comment by DigitVE** [comment]
Thank you, I'll post here logs tomorrow and ofc I used config `version: '3'` with latest version. `rpc.listen` in `.rr.yaml` looks fine and it works great on older versions than 2.8.0

**Comment by rustatian** [comment]
I would be happy to help you, but I can't reproduce this bug. Try to check the `CWD` as well, might be RR can't find the configuration (or try to use absolute path to the config). Might be `hosts` have some DNS values overwritten... And since this is `yaml`, check the indentation 😃

**Comment by rustatian** [comment]
This is 99.9% not an RR bug, so you should probably check your environment. It's hard to say why 2.8.0 works, but as far as I remember it was a bug with the `CWD` setting.

**Comment by DigitVE** [comment]
Thank you very much for your suggestions! We are still investigating this strange bug, but in the logs there are no any useful info after running these commands:

```
2023-06-19T07:26:50.650Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:26:50.542Z", "elapsed": "107.528082ms"}
2023-06-19T07:26:55.673Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:26:55.550Z", "elapsed": "122.983971ms"}
2023-06-19T07:27:00.650Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:00.541Z", "elapsed": "109.181682ms"}
2023-06-19T07:27:05.648Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:05.542Z", "elapsed": "106.036877ms"}
2023-06-19T07:27:10.648Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:10.541Z", "elapsed": "106.102353ms"}
2023-06-19T07:27:15.672Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:15.542Z", "elapsed": "130.248345ms"}
2023-06-19T07:27:20.667Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:20.542Z", "elapsed": "125.685295ms"}
2023-06-19T07:27:25.655Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:25.542Z", "elapsed": "113.458497ms"}
2023-06-19T07:27:30.651Z INFO http http log {"status": 200, "method": "GET", "URI": "http://localhost/readyz", "remote_address": "192.168.255.1", "start": "2023-06-19T07:27:30.541Z", "elapsed": "109.985258ms"}
```

Really appreciate for your ideas on this unreproducible problem 🙌

**Comment by rustatian** [comment]
The error itself is defined here: https://github.com/roadrunner-server/roadrunner/blob/master/internal/rpc/client.go#L74
It can only happen if your address was badly formed (e.g. `tcp:\127.0.0.1:8888`). Also, check the `CWD` where RR tries to find the configuration. Use `-w` (set working directory if necessary) or absolute path with `-c` to point to the configuration. Make sure the configuration exists and is well-formed.

**Comment by rustatian** [comment]
Few notes about the error: 
`dsn := strings.Split(addr, "://")` tries to split the string by the `://` well known net delimiter. So, the only way, you may have this error is badly formed configuration.

**Comment by rustatian** [comment]
Since I can't reproduce this bug and there are no steps on how to do so, I'm closing this issue, but you can continue the discussion (it's not locked).

**Comment by DigitVE** [comment]
I think the only possible way to resolve it somehow is to debug on prod 🥲 So, there is no way you can help me at this moment. Thank you very much for your comments, much appreciated

**Comment by FluffyDiscord** [comment]
I had the same issue. I am running RR in Docker and I set the container environment variables using Compose `env_file` for my app to use, and one of them was called `RR_RPC`. RR currently prefers this system environment variable instead of the one you set in your `.rr.yaml`. This confuses RR and on top of that theres something weird going on between Docker's network and RR.

TLDR: Make sure you don't have `RR_RPC` system environment variable set.

**Comment by rustatian** [comment]
Thanks @FluffyDiscord 👍 
Since this wans't documented I'll remove this behavior in the next minor release.
BTW, all envs starting with `RR_*` can entirely replace config values if match by name.

**Comment by idchlife** [comment]
For future lurkers. Error occurs also if you have invalid value anywhere in your configuration.

For example if you change relay to "somth" instead of "pipes" you will get this rpc error. Even if you have no rpc enabled.

So beware and check your config.

**Comment by rustatian** [comment]
@idchlife You can have RPC disabled. You don't have such an option. If you have RPC in your configuration, it is enabled.

**Comment by idchlife** [comment]
> [@idchlife](https://github.com/idchlife) You can have RPC disabled. You don't have such an option. If you have RPC in your configuration, it is enabled.

I don't think QUEUE/JOBS mode actually allows startup without RPC...

I mean I can't create Jobs class without providing RPC information. First argument should be provided - otherwise error occurs (I know documentation says you can use it without RPC, but unfortunately it is wrong, as of now).

Or maybe I am mistaken and that is not how you use Producer to push tasks into Queue (with creating Jobs and getting Queue with it)

And about error with configuration I mentioned above - my mistake. relay option expects "pipes" or actual tcp address for RPC. This is why I got RPC error when changed relay option to gibberish.
